[TensorExpr] Fuser: try merging adjacent fusion groups.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44756 [CI] Add profiling-te benchmarks.
* **#43671 [TensorExpr] Fuser: try merging adjacent fusion groups.**

Differential Revision: [D23360796](https://our.internmc.facebook.com/intern/diff/D23360796)

Commit:
[TensorExpr] Fuser: try merging adjacent fusion groups.

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Fuser: try merging adjacent fusion groups."


Differential Revision: [D23360796](https://our.internmc.facebook.com/intern/diff/D23360796)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Fuser: try merging adjacent fusion groups."


Differential Revision: [D23360796](https://our.internmc.facebook.com/intern/diff/D23360796)

[ghstack-poisoned]

eellison commented on test/cpp/tensorexpr/test_te_fuser_pass.cpp:
```diff
@@ -89,5 +89,30 @@ void testFuserPass_3() {
   }
 }
+
+void testFuserPass_4() {
```
nit: maybe `testFuserPass_merge_adjacent_fusion_groups`

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -379,11 +380,40 @@ class TensorExprFuser {
+      // If merging suceeded, save the merged group as the "previous" fusion
+      // group so that we can try to merge the next one into it.
+      if (prev_fusion_group) {
+        debugDumpFusionGroup(
```
small nit: theres a debug dump on each branch of the following if statement and a debug on the previous line, not sure how much information this adds

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -379,11 +380,40 @@ class TensorExprFuser {
+        debugDumpFusionGroup(
+            "Trying to merge into the previous fusion group: ",
+            prev_fusion_group);
+        if (canMerge(prev_fusion_group, fusion_group)) {
```
tryMerge can fail even when `canMerge` can fail, which is a little confusing. https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L443

Can we inline `canMerge` into tryMerge, and then call tryMerge here instead of `canMerge` ? (these are the only two callsites)

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -498,7 +528,7 @@ class TensorExprFuser {
 
     // Symbolic checks
-    REQ(canHandle(producer));
+    REQ(canHandle(producer) || producer->kind() == prim::TensorExprGroup);
```
could also add prim::TensorExprGroup to the symbols which we support

eellison reviewed (approved):
LGTM

Commit:
Update on "[TensorExpr] Fuser: try merging adjacent fusion groups."


Differential Revision: [D23360796](https://our.internmc.facebook.com/intern/diff/D23360796)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Fuser: try merging adjacent fusion groups."


Differential Revision: [D23360796](https://our.internmc.facebook.com/intern/diff/D23360796)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Fuser: try merging adjacent fusion groups."


Differential Revision: [D23360796](https://our.internmc.facebook.com/intern/diff/D23360796)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Fuser: try merging adjacent fusion groups."


Differential Revision: [D23360796](https://our.internmc.facebook.com/intern/diff/D23360796)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Fuser: try merging adjacent fusion groups."


Differential Revision: [D23360796](https://our.internmc.facebook.com/intern/diff/D23360796)

[ghstack-poisoned]
